File: src/main/java/com/zaid/patel/AI_Email_Assistant/Service/SupabaseAuthService.java
package com.zaid.patel.AI_Email_Assistant.Service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.zaid.patel.AI_Email_Assistant.DTo.AuthResponse;
import com.zaid.patel.AI_Email_Assistant.DTo.LoginRequest;
import com.zaid.patel.AI_Email_Assistant.DTo.SignUpRequest;
import com.zaid.patel.AI_Email_Assistant.Entity.Users;
import com.zaid.patel.AI_Email_Assistant.Repository.UsersRepository;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.*;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;

import java.util.Map;

@Service
public class SupabaseAuthService {

    private final WebClient webClient;
    private final UsersRepository usersRepository;
    private final ObjectMapper objectMapper;

    @Value("${supabase.url}")
    private String supabaseUrl;

    @Value("${supabase.anon.key}")
    private String anonKey;

    public SupabaseAuthService(WebClient.Builder webClientBuilder, UsersRepository usersRepository, ObjectMapper objectMapper) {
        this.webClient = webClientBuilder.build();
        this.usersRepository = usersRepository;
        this.objectMapper = objectMapper;
    }

    public AuthResponse signUp(SignUpRequest request) {
        Map<String, String> body = Map.of(
                "email", request.getEmail(),
                "password", request.getPassword()
        );

        String responseJson = webClient.post()
                .uri(supabaseUrl + "/auth/v1/signup")
                .header("apikey", anonKey)
                .header("Content-Type", "application/json")
                .bodyValue(body)
                .retrieve()
                .bodyToMono(String.class)
                .block();

        try {
            JsonNode root = objectMapper.readTree(responseJson);
            JsonNode userNode = root.path("user");
            String userId = userNode.path("id").asText();
            String email = userNode.path("email").asText();
            String accessToken = root.path("access_token").asText();
            String tokenType = root.path("token_type").asText();

            if (usersRepository.findByEmail(email).isEmpty()) {
                Users user = new Users(request.getName(), email, "");
                user.setId(java.util.UUID.fromString(userId));
                usersRepository.save(user);
            }

            return new AuthResponse(accessToken, tokenType, email, request.getName(), userId);
        } catch (Exception e) {
            throw new RuntimeException("Supabase signup failed: " + e.getMessage(), e);
        }
    }

    public AuthResponse signIn(LoginRequest request) {
        Map<String, String> body = Map.of(
                "email", request.getEmail(),
                "password", request.getPassword()
        );

        String responseJson = webClient.post()
                .uri(supabaseUrl + "/auth/v1/token?grant_type=password")
                .header("apikey", anonKey)
                .header("Content-Type", "application/json")
                .bodyValue(body)
                .retrieve()
                .bodyToMono(String.class)
                .block();

        try {
            JsonNode root = objectMapper.readTree(responseJson);
            JsonNode userNode = root.path("user");
            String userId = userNode.path("id").asText();
            String email = userNode.path("email").asText();
            String accessToken = root.path("access_token").asText();
            String tokenType = root.path("token_type").asText();

            String name = email;
            var localUser = usersRepository.findByEmail(email);
            if (localUser.isPresent()) {
                name = localUser.get().getName();
            }

            return new AuthResponse(accessToken, tokenType, email, name, userId);
        } catch (Exception e) {
            throw new RuntimeException("Supabase signin failed: " + e.getMessage(), e);
        }
    }

    public JsonNode getUser(String accessToken) {
        String responseJson = webClient.get()
                .uri(supabaseUrl + "/auth/v1/user")
                .header("Authorization", "Bearer " + accessToken)
                .header("apikey", anonKey)
                .retrieve()
                .bodyToMono(String.class)
                .block();

        try {
            return objectMapper.readTree(responseJson);
        } catch (Exception e) {
            throw new RuntimeException("Failed to get user: " + e.getMessage(), e);
        }
    }
}

File: apps/api/src/services/supabase-auth.service.js
import { env } from '../config/env.js'

const SUPABASE_AUTH_URL = `${env.SUPABASE_URL}/auth/v1`

async function supabaseFetch(path, options = {}) {
  const response = await fetch(`${SUPABASE_AUTH_URL}${path}`, {
    ...options,
    headers: {
      'apikey': env.SUPABASE_ANON_KEY,
      'Content-Type': 'application/json',
      ...options.headers,
    },
  })

  if (!response.ok) {
    const errBody = await response.text()
    throw new Error(errBody || `Supabase API error: ${response.status}`)
  }

  return response.json()
}

function mapAuthData(data, fallbackEmail) {
  const expiresAt =
    data.expires_at ??
    (data.expires_in ? Math.floor(Date.now() / 1000) + data.expires_in : null)

  return {
    accessToken: data.access_token,
    tokenType: data.token_type,
    email: data.user?.email || fallbackEmail,
    name: data.user?.user_metadata?.name || fallbackEmail,
    userId: data.user?.id,
    refreshToken: data.refresh_token,
    expiresAt,
  }
}

export async function signup(email, password, name) {
  const data = await supabaseFetch('/signup', {
    method: 'POST',
    body: JSON.stringify({ email, password }),
  })

  return mapAuthData(data, name || email)
}

export async function login(email, password) {
  const data = await supabaseFetch('/token?grant_type=password', {
    method: 'POST',
    body: JSON.stringify({ email, password }),
  })

  return mapAuthData(data, email)
}

export async function refresh(refreshToken) {
  const data = await supabaseFetch('/token?grant_type=refresh_token', {
    method: 'POST',
    body: JSON.stringify({ refresh_token: refreshToken }),
  })

  return mapAuthData(data, data.user?.email)
}

export async function getProfile(accessToken) {
  return supabaseFetch('/user', {
    headers: { 'Authorization': `Bearer ${accessToken}` },
  })
}
